internal/compiler-bridge/src/main/scala/xsbt/ExtractAPI.scala

```diff
@@ -198,33 +202,62 @@ class ExtractAPI[GlobalType <: Global](
   // The compiler only pickles static annotations, so only include these in the API.
   // This way, the API is not sensitive to whether we compiled from source or loaded from classfile.
   // (When looking at the sources we see all annotations, but when loading from classes we only see the pickled (static) ones.)
-  private def mkAnnotations(in: Symbol, as: List[AnnotationInfo]): Array[xsbti.api.Annotation] =
-    staticAnnotations(as).toArray.map { a =>
-      xsbti.api.Annotation.of(
-        processType(in, a.atp),
-        if (a.assocs.isEmpty)
-          Array(xsbti.api.AnnotationArgument.of("", a.args.mkString("(", ",", ")"))) // what else to do with a Tree?
-        else
-          a.assocs
-            .map {
-              case (name, value) =>
-                xsbti.api.AnnotationArgument.of(name.toString, value.toString)
-            }
-            .toArray[xsbti.api.AnnotationArgument]
-      )
-    }
+  private def mkAnnotations(in: Symbol, as: List[AnnotationInfo]): Array[xsbti.api.Annotation] = {
+    if (in == NoSymbol) ExtractAPI.emptyAnnotationArray
+    else
+      staticAnnotations(as) match {
+        case Nil => ExtractAPI.emptyAnnotationArray
+        case staticAs =>
+          staticAs.map { a =>
+            xsbti.api.Annotation.of(
+              processType(in, a.atp),
+              if (a.assocs.isEmpty)
+                Array(xsbti.api.AnnotationArgument.of("", a.args.mkString("(", ",", ")"))) // what else to do with a Tree?
+              else
+                a.assocs
+                  .map {
+                    case (name, value) =>
+                      xsbti.api.AnnotationArgument.of(name.toString, value.toString)
+                  }
+                  .toArray[xsbti.api.AnnotationArgument]
+            )
+          }.toArray
+      }
+  }
 
-  private def annotations(in: Symbol, s: Symbol): Array[xsbti.api.Annotation] =
-    enteringPhase(currentRun.typerPhase) {
+  // HOT method, hand optimized to reduce allocations and needless creation of Names with calls to getterIn/setterIn
+  // on non-fields.
+  private def annotations(in: Symbol, s: Symbol): Array[xsbti.api.Annotation] = {
+    val saved = phase
+    phase = currentRun.typerPhase
+    try {
       val base = if (s.hasFlag(Flags.ACCESSOR)) s.accessed else NoSymbol
       val b = if (base == NoSymbol) s else base
       // annotations from bean methods are not handled because:
       //  a) they are recorded as normal source methods anyway
       //  b) there is no way to distinguish them from user-defined methods
-      val associated =
-        List(b, b.getterIn(b.enclClass), b.setterIn(b.enclClass)).filter(_ != NoSymbol)
-      associated.flatMap(ss => mkAnnotations(in, ss.annotations)).distinct.toArray
+      if (b.hasGetter) {
```

Comment:
This change needs the most careful review. We should probably build a version that has the old and new implementations in place that asserts they agree.

Reply:
Looks correct under the assumption that the compiler does not synthesize setters if there are no getters. We have a biggish community build in bloop (https://github.com/scalacenter/bloop/tree/master/build-integrations), I can try out the optimized Zinc version when this PR is ready and we assert that we don't break incremental compilation for existing projects.

Reply:
Is it possible to write out common cases as tests in https://github.com/sbt/zinc/blob/1.1.x/internal/compiler-bridge-test/src/test/scala/xsbt/ExtractAPISpecification.scala?